[OpenTelemetry] Add OnEnding span processor functionality (#6617)
open-telemetry/opentelemetry-dotnet: docs/diagnostics/experimental-apis/OTEL1005.md

# OpenTelemetry .NET Diagnostic: OTEL1004

## Overview

> This is an experimental API for modifying spans before they end/close

Reviewer: ```suggestion
This is an experimental API for modifying spans before they end/close.
```

### Details

#### ExtendedBaseProcessor

The abstract class `ExtendedBaseProcessor` provides an extension of the

> `BaseProcessor` that allows spans to be modified before they end as per the

Reviewer: ```suggestion
`BaseProcessor` class that allows spans to be modified before they end as per the
```

OpenTelemetry specification. It provides the `OnEnding` function that is called
during the span `End()` operation. The end timestamp MUST have been computed
(the `OnEnding` method duration is not included in the span duration). The Span
object MUST still be mutable (i.e., `SetAttribute`, `AddLink`, `AddEvent` can be
called) while `OnEnding` is called. This method MUST be called synchronously
within the [`Span.End()` API](api.md#end), therefore it should not block or

> throw an exception. If multiple `SpanProcessors` are registered, their

Reviewer: ```suggestion
throw an exception. If multiple `SpanProcessor`s are registered, their
```
